drv/stm32xx-gpio-common/src/server.rs

```diff
@@ -0,0 +1,278 @@
+                compiler_error!("unsupported or missing SoC model feature");
+            }
+        }
+    } else if #[cfg(feature = "family-stm32h7")] {
+        use stm32h7 as pac;
+        cfg_if::cfg_if! {
+            if #[cfg(feature = "model-stm32h743")] {
+                use pac::stm32h743 as device;
+            } else if #[cfg(feature = "model-stm32h753")] {
+                use pac::stm32h753 as device;
+            } else {
+                compiler_error!("unsupported or missing SoC model feature");
+            }
+        }
+    } else {
+        compiler_error!("unsupported or missing SoC family feature");
+    }
+}
+
+/// Returns a reference to the GPIO peripheral corresponding to `port`. Because
+/// GPIO peripherals tend to have many different types, this hides the true type
+/// behind a `&dyn AnyGpioPeriph`.
+///
+/// # Safety
+///
+/// I guess this could be used unsafely if you used it within a system using the
+/// Embedded HAL which relies on uniqueness of peripheral access? In practice,
+/// as long as you ensure you're not doing this, you're probably ok.
+#[cfg(feature = "server-support")]
+pub unsafe fn get_gpio_regs(port: Port) -> &'static dyn AnyGpioPeriph {
+    match port {
+        Port::A => &*device::GPIOA::ptr(),
+        Port::B => &*device::GPIOB::ptr(),
+        Port::C => &*device::GPIOC::ptr(),
+        Port::D => &*device::GPIOD::ptr(),
+
+        #[cfg(feature = "has-port-gpioe")]
+        Port::E => &*device::GPIOE::ptr(),
+        #[cfg(feature = "has-port-gpiof")]
+        Port::F => &*device::GPIOF::ptr(),
+        #[cfg(feature = "has-port-gpiog")]
+        Port::G => &*device::GPIOG::ptr(),
+        #[cfg(feature = "has-port-gpioh")]
+        Port::H => &*device::GPIOH::ptr(),
+        #[cfg(feature = "has-port-gpioi")]
+        Port::I => &*device::GPIOI::ptr(),
+        #[cfg(feature = "has-port-gpioj")]
+        Port::J => &*device::GPIOJ::ptr(),
+        #[cfg(feature = "has-port-gpiok")]
+        Port::K => &*device::GPIOK::ptr(),
+    }
+}
+
+/// A GPIO peripheral that can be manipulated in abstract ways without knowing
+/// its type.
+#[cfg(feature = "server-support")]
+pub trait AnyGpioPeriph {
+    fn configure(&self, pins: u16, atts: u16);
+    fn set_reset(&self, set: u16, reset: u16);
+    fn toggle(&self, pins: u16);
+    fn read(&self) -> u16;
+}
+
+/// Adapter from `GpioPeriph`, the trait implemented for disjoint GPIO
+/// peripheral types, to the general `AnyGpioPeriph`.
+#[cfg(feature = "server-support")]
+impl<T: GpioPeriph> AnyGpioPeriph for T {
+    fn configure(&self, pins: u16, atts: u16) {
+        // The GPIO config registers come in 1, 2, and 4-bit per field variants.
+        // The user-submitted mask is already correct for the 1-bit fields; we
+        // need to expand it into corresponding 2- and 4-bit masks. We use an
+        // outer perfect shuffle operation for this, which interleaves zeroes
+        // from the top 16 bits into the bottom 16.
```

Comment:
I love the use of perfect shuffling here!

Reply:
Heh, thanks. The Hacker's Delight outer perfect shuffle routine is also _ridiculously_ cheap on ARMv7-M because the constants happen to match patterns that can be materialized cheaply by Thumb-2 instructions. It is _much worse_ on ARMv6-M, since those particular instructions were among the ones removed to simplify the architecture. But, ARMv6-M takes more instructions to do _everything,_ so I'm not terribly concerned.